### ISSUE TITLE
Homebrew/framework install scheme not honored for scripts

### ISSUE THREAD

**Issue by alexreg** [pypa/distutils#152]
### setuptools version

60.10.0

### Python version

3.9.12

### OS

macOS 12.3.1

### Additional environment information

Uses `python` formula from macOS Homebrew.

### Description

Installing a package with console scripts defined using `python setup.py develop` does not actually install the console scripts. (It works with `python setup.py install`, however.)

This issue was previously discussed on the [pypa/pip repo](https://github.com/pypa/pip/issues/11067).

### Expected behavior

Setuptools should install console script in the appropriate place.

### How to Reproduce

1. Create the following files.

`setup.py`
```python
from setuptools import setup

setup(
    entry_points = {
        'console_scripts': [
            'foo = foopackage:main',
        ],
    }
)
```

`foopackage/__init__.py`
```python
def main():
    print("foo")
```

2. Run `python setup.py develop`.

### Output

```console
/opt/homebrew/lib/python3.9/site-packages/setuptools/command/easy_install.py:144: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/opt/homebrew/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
running egg_info
creating foopackage.egg-info
writing foopackage.egg-info/PKG-INFO
writing dependency_links to foopackage.egg-info/dependency_links.txt
writing entry points to foopackage.egg-info/entry_points.txt
writing top-level names to foopackage.egg-info/top_level.txt
writing manifest file 'foopackage.egg-info/SOURCES.txt'
listing git files failed - pretending there aren't any
reading manifest file 'foopackage.egg-info/SOURCES.txt'
writing manifest file 'foopackage.egg-info/SOURCES.txt'
running build_ext
Creating /opt/homebrew/lib/python3.9/site-packages/foopackage.egg-link (link to .)
foopackage 0.0.0 is already the active version in easy-install.pth
Installing foo script to /opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/bin

Installed /Users/alex/foo
Processing dependencies for foopackage==0.0.0
Finished processing dependencies for foopackage==0.0.0
```

**Comment by jaraco** [comment]
I see now that the scheme is essentially irrelevant, because it's the `prefix` / `install_base` that's being customized by the `pydistutils.cfg` file.

**Comment by jaraco** [comment]
Trying to figure out how `sysconfig` resolves scripts to `/opt/homebrew`, I took a look at the `base` config var, but that's not where it happens:

```
distutils main $ py -c "import sysconfig; print(sysconfig.get_config_var('base')); print(sysconfig.get_path('scripts'))"
/opt/homebrew/opt/python@3.10/Frameworks/Python.framework/Versions/3.10
/opt/homebrew/bin
```

**Comment by abravalheri** [comment]
Hi @jaraco, I remember seeing that different Python versions distributed by homebrew would use different techniques to customize the installation scheme.

The customization for Python3.10 is likely to use a different approach than the ones for 3.9, 3.8, 3.7.

Let me see if I can find any links...

**Comment by jaraco** [comment]
Oh! Good tip. So maybe this issue doesn't affect Python 3.10?

**Comment by jaraco** [comment]
I think maybe that's true:

```
distutils main $ py -c "import distutils.command.install as inst; print(inst._load_scheme(inst._resolve_scheme('posix_prefix'))['scripts'])"
/opt/homebrew/bin
```

**Comment by abravalheri** [comment]
Yeah, I think it does not affect 3.10. However it will still affect older versions (with EOS programmed to October 2025)...

I don't know if the Homebrew team is planning to keep supporting these versions and if they plan to change the customization mechanism for future patch releases of the "bottles"...

**Comment by jaraco** [comment]
In that case, Python 3.9 and earlier will probably not get the install schemes treatment, so distutils is left without the patched context left by Homebrew. I'm now starting to agree with abravalheri's recommendation to honor pydistutils.cfg if found in the system site packages. In addition to limiting it to framework builds, I'd also want to limit it to Python 3.9 and earlier.

Another option could be simply to declare this use case as unsupported. It seems to be pretty rare. I believe there's a workaround where if you run with `SETUPTOOLS_USE_DISTUTILS=stdlib` set in the environment, that should cause the old distutils to be invoked at the stdlib location, and thus honor the `pydistutils.cfg`. e.g.:

```
$ env SETUPTOOLS_USE_DISTUTILS=stdlib pip install -e .
```

@alexreg Could you try that and confirm if the scripts install as expected?

**Comment by alexreg** [comment]
@jaraco Yep, that works too.

I should note, however, that subsequently doing `pip uninstall [package]` does *not* remove the console script, only the package itself. This was likewise the case with the other approach (symlinking `distutils.cfg`).

**Comment by jaraco** [comment]
> Yep, that works too.

That's great news. Now the follow-up question - will this workaround be sufficient to get by until Python 3.9 is sunset or do you need a more robust solution?

Also, I should note that setuptools is working on a PEP 660 based solution that should obviate the need for pip to call setup.py develop and should address the issue, even for older Pythons. It just needs to land.

> I should note, however, that subsequently doing `pip uninstall [package]` does _not_ remove the console script, only the package itself. This was likewise the case with the other approach (symlinking `distutils.cfg`).

I suspect that's the case irrespective of python versions or setuptools versions. I'm pretty sure pip relies on setuptools' setup.py develop to uninstall and it doesn't remove any scripts. That's a bug that's not going to get fixed until Setuptools adopts a PEP 660 solution.

**Comment by jaraco** [comment]
In #155, I've drafted a compatibility shim that should enable installing resources to the homebrew prefix. I've tested the functionality superficially, but haven't tested it in a real-world environment. Here's how you can test it in your environment:

1. Install a late (or the latest) version of Setuptools.
2. Override `setuptools._distutils` with this copy of distutils.

```
$ wget -O - https://github.com/pypa/distutils/archive/refs/heads/refactor/142-allow-pathlike.zip | bsdtar -x -C $(py -3.9 -c "import distutils, os; print(os.path.dirname(distutils.__file__))") --strip-components 2 --exclude docs/
```

Note the command above requires `wget`, `bsdtar`, and `python-launcher` (all available in homebrew). Note also that it applies the code directly to your environment, so use at your own risk. If you'd rather take it safer, just download the code yourself and copy over the files to your site-packages/setuptools/_distutils directory manually.

3. Install as editable something that installs scripts and confirm that they're installed to $homebrew_prefix/bin as expected.

**Comment by alexreg** [comment]
@jaraco Thank you, that's great. A brief test suggests your solution works for me too. (As before, `pip uninstall` does not work, but that's a small matter in comparison. Shouldn't be a problem to wait for an implementation of PEP 660.)

**Comment by jaraco** [comment]
Thanks Alexander for reviewing. I'll merge it and get it out to Setuptools and we can iterate if needed.

**Comment by Bo98** [comment]
Just to clarify: patching `sysconfig` for <= 3.9 on the Homebrew side was (and still is) considered as an option. There is however one big breakage (which we accepted as a "breaking change" for 3.10): `--prefix` no longer works properly for site-packages etc. This is because that works by swapping `base` and `base` is not used in the modified `sysconfig` for some items. We did make quite a bit of use of `--prefix` in dependents, so some migration work would be required to stop using it (which was easier to cover in a 3.9 -> 3.10 transition).

This is a limitation in `sysconfig`. Fedora also share the same issue: https://bugzilla.redhat.com/show_bug.cgi?id=2026979. There's a few areas of `sysconfig` that really haven't matured enough yet, e.g. https://github.com/python/cpython/issues/88142.

The reason `--prefix` works in `pip` is because we (ab)use this hack: https://github.com/pypa/pip/blob/ae324d17033253626725330c6635c0089927fbf5/src/pip/_internal/locations/_sysconfig.py#L157-L161. This hack wasn't actually Homebrew-specific. It was specific to Apple's Python and we just decided to name our scheme the same to make use of that.
